### ISSUE TITLE
.NET 4.5 console app bundled with mono 6.4.0 cannot be signed, or crashes when run

### ISSUE THREAD

**Issue by DancingDev** [mono/mono#17881]
I've been struggling with this for a month now, unable to bundle a .NET exe with mono such that it runs correctly on a target macOS 10.14 after having been codesignged with strict and hardened runtime options.  Either working bundle cannot be signed, or I have a non-working bundle that can be signed.

Please see for test project used to reproduce the issue.  The app is a simple "hello world" .NET console app with no external dependencies.
https://github.com/DancingDev/Sign-.NET-Console-App-bundled-with-Mono

I'll repeat here:

Related bug, cannot find any other references to this issue on mono github / stack overflow / xamarin community forums
https://xamarin.github.io/bugzilla-archives/52/52443/bug.html

Requirements:
   - Mac OS 10.14.6
   - Visual Studio for Mac 8.3.9
   - XCode 11
   - Mono 6.4.0
   - A Signing Certificate (self signed used below)
      - https://support.apple.com/guide/keychain-access/create-self-signed-certificates-kyca8916/mac

Steps to reproduce issue(s)

Case 1:
1. compile the testConsoleAppProduct in VS with release config, produces testConsoleApp.exe
2. in the project folder execute './bundleTestAppWithMono.sh simple' which calls mkbundle with --simple flag
   - https://www.mono-project.com/docs/tools+libraries/tools/mkbundle/ per documentation here
   - produces testConsoleApp binary which works when run on the command line "./bin/Release/testConsoleApp"
```
mkbundle -v -o testConsoleApp --deps --simple testConsoleApp.exe --sdk $MONOROOT/Versions/Current
```
   
3. in the project folder execute './signTestConsoleApp.sh strict {cert-to-use}', which code signs with --strict -o runtime and entitlements
```
codesign -f --strict --verbose --entitlements $ENTSFILE -o runtime -s "$CERT" testConsoleApp --timestamp
```

Expected: signing to be successful

Actual: signing is unsuccessful, 'testConsoleApp: main executable failed strict validation'

Case 2:
1. compile the testConsoleAppProduct in VS with release config, produces testConsoleApp.exe
2. in the project folder execute './bundleTestAppWithMono.sh L' which calls mkbundle with -L and passes mono directory for the path
   - https://www.mono-project.com/docs/tools+libraries/tools/mkbundle/ per documentation here
   - produces testConsoleApp binary which seg faults, error code 11, when run on the command line "./bin/Release/testConsoleApp"
   - stack trace here: https://github.com/DancingDev/Sign-.NET-Console-App-bundled-with-Mono/blob/master/testConsoleApp/testConsoleApp-Crash-Trace.txt
```
 mkbundle -v -o testConsoleApp --deps testConsoleApp.exe -L $MONOROOT/Versions/Current/lib/mono/4.5/
```

3. in the project folder execute './signTestConsoleApp.sh strict {cert-to-use}', which code signs with --strict -o runtime and entitlements

Expected: signing to be successful

Actual: signing is successful, but the program still seg faults when executed "./bin/Debug/testConsoleApp"

## On which platforms did you notice this

[X ] macOS
[ ] Linux
[ ] Windows

## Stacktrace

https://github.com/DancingDev/Sign-.NET-Console-App-bundled-with-Mono/blob/master/testConsoleApp/testConsoleApp-Crash-Trace.txt

**Comment by HinTak** [comment]
See also https://github.com/HinTak/Font-Validator/issues/53 , in particular on how mkbundle works after mono 4.8.x

**Comment by DancingDev** [comment]
For anyone looking for updates here.  I was able to find a solution that no longer uses mkbundle by creating a new Xamarin App that wraps my console app.  

The way Xamarin compiles and bundles .NET applications produces an assembly that can be codesigned.
https://docs.microsoft.com/en-us/xamarin/mac/internals/architecture

My solution does require access to the source code of console app.

I updated my post here: https://forums.xamarin.com/discussion/174937/how-to-bundle-a-net-app-with-mono-such-that-it-can-be-signed-or-such-that-it-can-run-w-o-crashing

**Comment by HinTak** [comment]
I have written a python script using macholib, based on the advice at the bottom of https://github.com/pyinstaller/pyinstaller/wiki/Recipe-OSX-Code-Signing . The fixed binary is detailed in https://github.com/HinTak/Font-Validator/issues/53 for testing. The nice thing about this approach is that it is a post-release fix, no need to rebuild the whole thing - just run the script on the older binary to get a new one.

**Comment by HinTak** [comment]
Here is the script -
[mono-codesign-fix.py.txt](https://github.com/mono/mono/files/4450266/mono-codesign-fix.py.txt)
Ideally the same functionality should be integrated to mkbundle. Then if the loader is patched according to the patch I posted in https://github.com/mono/mono/issues/18826 , it should work.

**Comment by HinTak** [comment]
The script is posted in https://github.com/HinTak/mono-modification , together with patched binaries to fix https://github.com/mono/mono/issues/18826 .

**Comment by ylatuya** [comment]
This issue with the creation of bundles also breaks others scenarios, such as trying to add a new rpath with install_name_tool

```
$ install_name_tool -add_rpath "@loader_path" LongoMatch                                                                         
$ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: the __LINKEDIT segment does not cover the end of the file (can't be processed) in: LongoMatch
```

The problem seems to be that the __LINKEDIT segment coming from the mono runtime binary used to create the new binary is not updated after creating the bundle with the new size of the binary

This is the output from `otool -l LongoMatch`

```
Load command 3
      cmd LC_SEGMENT_64
  cmdsize 72
  segname __LINKEDIT
   vmaddr 0x000000010041a000
   vmsize 0x000000000008e000
  fileoff 4190208
 filesize 500492
  maxprot 0x00000007
 initprot 0x00000001
   nsects 0
    flags 0x0
```

fileoff + filesize should be equal to the new length of the file including the embedded assembly deps.

**Comment by HinTak** [comment]
@ylatuya - yes, the issue is all well-understood. That's what my script does - run it just before you sign.

**Comment by ylatuya** [comment]
@HinTak  I am using your mono-codesign-fix script with the mono binary built from sources using your fix https://github.com/mono/mono/pull/19580 for mkbundle. When built from sources, after running the script, codesign fails again with  "main executable failed strict validation".
Here is the output of the script:
```
entry(LC_SYMTAB):
	 LC_SYMTAB <symtab_command symoff=4199952 nsyms=42617 stroff=4884700 strsize=596208> b''
64988594 5480908
4th(LC_SEGMENT_64/__LINKEDIT):
	 LC_SEGMENT_64 <segment_command_64 segname=b'__DATA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' vmaddr=4299100160 vmsize=126976 fileoff=4132864 filesize=12288 maxprot=3 initprot=3 nsects=4 flags=0> [<section_64 sectname=b'__la_symbol_ptr\x00' segname=b'__DATA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' addr=4299100160 size=2752 offset=4132864 align=3 reloff=0 nreloc=0 flags=7 reserved1=375 reserved2=0 reserved3=0>, <section_64 sectname=b'__data\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' segname=b'__DATA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' addr=4299102912 size=9504 offset=4135616 align=4 reloff=0 nreloc=0 flags=0 reserved1=0 reserved2=0 reserved3=0>, <section_64 sectname=b'__bss\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' segname=b'__DATA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' addr=4299112448 size=91544 offset=0 align=6 reloff=0 nreloc=0 flags=1 reserved1=0 reserved2=0 reserved3=0>, <section_64 sectname=b'__common\x00\x00\x00\x00\x00\x00\x00\x00' segname=b'__DATA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' addr=4299204000 size=20080 offset=0 align=4 reloff=0 nreloc=0 flags=1 reserved1=0 reserved2=0 reserved3=0>]
64988594 4145152 False
```

Using mono from the installer the output is:
```
entry(LC_SYMTAB):
	 LC_SYMTAB <symtab_command symoff=4243664 nsyms=9758 stroff=4402724 strsize=288144> b''
63890948 4690868
4th(LC_SEGMENT_64/__LINKEDIT):
	 LC_SEGMENT_64 <segment_command_64 segname=b'__LINKEDIT\x00\x00\x00\x00\x00\x00' vmaddr=4299268096 vmsize=581632 fileoff=4190208 filesize=500660 maxprot=7 initprot=1 nsects=0 flags=0> []
63890948 4690868 False
```

**Comment by ylatuya** [comment]
@HinTak The script is assuming the __LINKEDIT is in the 4th segment but that's not always true.
I had to modify the script to do:
```
 linkedit = [c for c in h.commands if \
        hasattr(c[1], 'segname') and \
        str(c[1].segname).startswith("b'__LINKEDIT")][0]
```

**Comment by HinTak** [comment]
Yes the script is limited and only tested on the mono binaries I built ( https://github.com/HinTak/mono-modification ) .

The latest Fontval for mac is basically done with:

- replace `/bin/mono` in the fetch-target of mkbundle with mine from ( https://github.com/HinTak/mono-modification - use a matching version, the sgen one, as the host's)

- run `mkbundle --cross ...` as usual.

- run the python adjustment script after that.

It is quite likely you don't build mono as I did, so the script does not quite work for you.

@ylatuya can you zip up and  attach one of you mono binaries to this?

**Comment by ylatuya** [comment]
I built mono from sources (sh autogen.sh  && make && make install). In that binary the __LINKEDIT segment is not the 4th in the commands list. Using the solution I proposed in my previous comment that looks for the segment name instead of hardcoding the position fixes the problem.

**Comment by HinTak** [comment]
Can I have your binary to test it? No, I did not do `make install` - force of habbit :-). I was cross-compiling (so `make install` is not appropriate), and my cross-compiler is only single arch (x86_64) so does not generate fat binaries - but the official fetch-target is also single-arch. And also somehow I know it is just `mono/mini` I need from the build directory...

**Comment by HinTak** [comment]
@ylatuya thanks for the code change, committed.

**Comment by HinTak** [comment]
@ylatuya Argh, I see LongoMatch is a proprietary/commercial product. Fair enough. I'd like to know in what way your build differs from the official/mine, though. Can you tell me what's the first 4 segment names? I have __PAGEZERO, __TEXT, __DATA, __LINKEDIT ... Somehow you must have another one before __LINKEDIT, I think.

**Comment by CoffeeFlux** [comment]
Now that #19580 is merged, we just need to get the updated signing functionality into mkbundle.

**Comment by nor0x** [comment]
@CoffeeFlux any update on this? i'm currently fighting with macOS notarization because of this issue. 

I have a Xamarin.Mac app which includes an executable (built with `mkbundle`) in `Contents/Resources`. the notarization process fails with `The signature of the binary is invalid.` and `codesign` gives me the following message `fatal error: the __LINKEDIT segment does not cover the end of the file (can't be processed) in: ...`.

I'm running the python script with the modifications by @ylatuya but it doesn't seem to modify the binary since the output is `The binary was signed. Not modifying.` Also `otool` shows that there is already a `LC_CODE_SIGNATURE` in the binary.

This is the mkbundle command i'm using to build the executable:
` mkbundle -v --simple -L exe1.exe dll1.dll dll2.dll -z --static --deps --cross /Library/Frameworks/Mono.framework/Versions/Current -o MyOutput`

**Comment by nor0x** [comment]
@HinTak thanks for the reply. I'm currently on 6.12 because i thought the fix was backported. I will try with 6.14 👍

**Comment by HinTak** [comment]
@nor0x I haven't been monitoring whether the change has been backported; all I am sure about is that the relevant change was merged into 6.13.x (mono follows the older Linux-like practice where odd number minor releases are dev releases for the next even number ones) - so 6.14+ is the stable release to use, if you don't want to build mono youself.

**Comment by HinTak** [comment]
@nor0x Argh, you are correct - it was backported to some late 6.12.xx . I'll need to check what xx is, if you really want to know.

**Comment by nor0x** [comment]
i'm using `6.12.0.140` from the latest Visual Studio Mac Preview

**Comment by HinTak** [comment]
@nor0x Yes, part of the fix is in 6.12.0.98 onwards. I am curious at what stage did you see  "codesign gives me the following message fatal error: the __LINKEDIT segment does not cover the end of the file (can't be processed) in: ...." though? what does `mkbundle --local-targets` and `mono --version` says?

**Comment by nor0x** [comment]
`fatal error: the __LINKEDIT segment does not cover the end of the file (can't be processed) in: ...`
is the result of the following codesign command which runs after the python script.

`codesign -v --force --no-strict -o runtime --timestamp --sign MYSIGNINGID --entitlements my.entitlements MyOutput`


this is the output of `mono --version`:
```
Mono JIT compiler version 6.12.0.140 (2020-02/51d876a041e Thu Apr 29 10:44:55 EDT 2021)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```
here is `mkbundle --local-targets`:
```
Available targets locally:
	default	- Current System Mono
```

**Comment by HinTak** [comment]
@nor0x okay, it appears that your binary is not quite what the script can handle. If it is something confidential you might want to send it by email privately? Otherwise you'll need to fix the script yourself...

**Comment by nor0x** [comment]
sorry for the delay @HinTak . I don't want to post the binary here publicly on GitHub but i would send it to you via email if that's okay?

thank you

**Comment by HinTak** [comment]
@nor0x I understand. E-mail as in most commits, in particular, https://github.com/HinTak/mono-modification/blob/0a73e37c586019cc987e48b18ca38ce0a21a5d4b/mono-6.12.0.98/0001-2020-02-Issue18826-fix-20358.patch#L82

**Comment by HinTak** [comment]
Thanks to @nor0x uncovers an interesting issue using mkbundle on Mac OS X to bundle for Mac OS X ie. building your app on mac with mono for mac without mono. (I myself have always been cross-compiling from Linux for mac...). One needs to strip the signature first before mkbundle, fix and sign. The codesign fix script has been updated to give an appropriate message.

cc @roslynn
